### DISCUSSION TITLE
v2.0.0: Custom model usage and cost tracking

### DISCUSSION THREAD

**Discussion by marcklingen** [langfuse/langfuse#998] [Announcements]
> [!IMPORTANT]
> This is a major release of Langfuse as it requires to run a [one-off (non-blocking) migration script](https://langfuse.com/changelog/2024-01-29-custom-model-prices#upgrade-path) on historical data to ensure accurate LLM costs in Langfuse.
> - No changes are required if you use Langfuse Cloud or are only concerned about newly ingested traces.
> - There are no other breaking changes.

## What's Changed
* **Fully rebuilt usage/cost tracking of LLM generations in Langfuse** by @maxdeichmann and @marcklingen 
  * New _model definition_ abstraction which allows to:
     * Quickly support newly emerging models.
     * Track changes in model prices over time.
     * Add custom models/prices at the project level.
  * You can also set (token) usage and cost via the API when ingesting traces.
    * If you want to set cost via the SDKs, make sure to upgrade to [JS 2.4.0](https://github.com/langfuse/langfuse-js/releases/tag/v2.4.0) and [Python 2.7.3](https://github.com/langfuse/langfuse-python/releases/tag/v2.7.3) or higher.
  * Usage and cost information on all UI tables and APIs
  * See [changelog](https://langfuse.com/changelog/2024-01-29-custom-model-prices) and [docs](https://langfuse.com/docs/model-usage-and-cost) for full details on this change.
* Other changes
  * feat(ui): add improved prompt ui with better versioning by @RichardKruemmel in https://github.com/langfuse/langfuse/pull/961
  * feat(ui): filtering by scores to the generations table in https://github.com/langfuse/langfuse/pull/880
  * feat: add new openai model prices by @maxdeichmann in https://github.com/langfuse/langfuse/pull/967
  * feat(db): improve seeding script for prompts by @RichardKruemmel in https://github.com/langfuse/langfuse/pull/965
  * fix(security): add validation for html in inputs by @RichardKruemmel in https://github.com/langfuse/langfuse/pull/927
  * fix(sessions): composite ids by @marcklingen in https://github.com/langfuse/langfuse/pull/946
  * fix(ui): retrieve prompts created via API by @maxdeichmann in https://github.com/langfuse/langfuse/pull/963
  * fix: error log for api prompts by @RichardKruemmel in https://github.com/langfuse/langfuse/pull/966
  * chore(deps): bump dependencies in https://github.com/langfuse/langfuse/pull/970, https://github.com/langfuse/langfuse/pull/936, https://github.com/langfuse/langfuse/pull/979, https://github.com/langfuse/langfuse/pull/986, https://github.com/langfuse/langfuse/pull/983, https://github.com/langfuse/langfuse/pull/985
  * refactor: typing of AutoCompleteOption by @marcklingen in https://github.com/langfuse/langfuse/pull/978
  * refactor: db logging in dev mode by @maxdeichmann in https://github.com/langfuse/langfuse/pull/952

**Full Changelog**: https://github.com/langfuse/langfuse/compare/v1.34.1...v2.0.0

<hr /><em>This discussion was created from the release <a href='https://github.com/langfuse/langfuse/releases/tag/v2.0.0'>v2.0.0</a>.</em>

**Comment by marcklingen** [comment]
We are very excited about this release as it makes usage and cost tracking super flexible and customizable.

For full details, check out the [changelog](https://langfuse.com/changelog/2024-01-29-custom-model-prices) and [docs](https://langfuse.com/docs/model-usage-and-cost) page.

**Why is this a major version when it is generally not breaking?** We made this a new major version because we know some teams use Langfuse's usage data for rate-limiting or billing. We wanted to avoid accidentally breaking this logic before these teams ran the migration script for historical data.

Looking forward to any feedback/ideas regarding cost/usage tracking in Langfuse!

**Comment by ghost** [comment]
Wasn't able to send cost via langfuse v`2.2.1` python

Is this live on the cloud app yet? @marcklingen

**Reply by ghost** [reply]
same issue on v`2.7.1`

**Reply by ghost** [reply]
code: 

```python
print(f"kwargs response cost: {kwargs['response_cost']}")
        trace.generation(
            name=metadata.get("generation_name", "litellm-completion"),
            id=metadata.get("generation_id", None),
            startTime=start_time,
            endTime=end_time,
            model=kwargs["model"],
            modelParameters=optional_params,
            input=input,
            output=output,
            usage={
                "prompt_tokens": response_obj["usage"]["prompt_tokens"],
                "completion_tokens": response_obj["usage"]["completion_tokens"],
                "input_cost": kwargs["response_cost"],
                "output_cost": kwargs["response_cost"],
            },
            metadata=metadata,
        )
```

**Reply by maxdeichmann** [reply]
@krrishdholakia looking into this now.

**Reply by maxdeichmann** [reply]
@krrishdholakia this only works as of `2.7.3` in langfuse python. I created a litellm PR (https://github.com/BerriAI/litellm/pull/1704) with questions.
